contracts/Nexus.sol

```diff
@@ -237,6 +237,20 @@ contract Nexus is INexus, EIP712, BaseAccount, ExecutionHelper, ModuleManager, U
         return IValidator(validator).isValidSignatureWithSender(msg.sender, computeHash, truncatedSignature);
     }
 
+    /// @notice Retrieves the address of the current implementation from the EIP-1967 slot.
+    /// @notice Checks the 1967 implementation slot, if not found then checks the slot defined by address (Biconomy V2 smart account)
+    /// @return implementation The address of the current contract implementation.
+    function getImplementation() external view returns (address implementation) {
+        assembly {
+            implementation := sload(_ERC1967_IMPLEMENTATION_SLOT)
+        }
+        if(implementation == address(0)) {
+            assembly {
+            implementation := sload(address())
```

Comment:
what is this magic?

Reply:
which line?

Reply:
`address()` opcode returns the same as `address(this)` 
so the contract's address is used as number of the storage slot (it is converted to bytes32) to store the address of the implementation

so this line loads it from storage using the opcode to obtain the slot where the implementation address is stored.

Reply:
yup,
https://github.com/bcnmy/scw-contracts/blob/main/contracts/smart-account/Proxy.sol#L23